### ISSUE TITLE
serialize mutates the given schema

### ISSUE THREAD

**Issue by lfabreges** [feathersjs/feathers-hooks-common#158]
### Steps to reproduce

1. Create a schema function:
```js
const schema = hook => {
  console.log("called");
  return {};
};
```

2. Add a serialize hook:
```js
hooks.serialize(schema)
```

3. Call the service at least two times

### Expected behavior
The schema function should be called for every service call and the schema constant should not be mutated by the serialize hook call.

### Actual behavior
The schema function is called only one time, then replaced with the returned object, the problem seems to be in the serialize function that mutate the given schema with this line:
```js
schema = typeof schema === 'function' ? schema(hook) : schema;
```

### System configuration

Tell us about the applicable parts of your setup.

**Module versions**
3.0.0

**NodeJS version**:
7.8.0

**Operating System**:
MacOSX

**React Native Version**:
0.44 RC 0

**Comment by eddyystop** [comment]
What DB, what adapter are you using?

If its feathers-memory or feathers-nedb with `inMemory:true` then you have to clone the result before calling populate or serialize. We are trying to figure out what to do with this undocumented gotcha.

**Comment by eddyystop** [comment]
Thanks for the catch.

**Comment by eddyystop** [comment]
I was unable to reproduce the problem with the following
```js
  it('schema param is not changed', () => {
    const schema1 = hook => {
      console.log('shouldBe called');
      return {};
    };
  
    const hook = clone(hookAfter);
    hook.result = { userId: 'as61389dadhga62343hads6712',
      postId: 1,
      updatedAt: 1480793101475,
      _include: [ 'post' ],
      _elapsed: { post: 1, total: 1 },
      post:
      { id: 1,
        title: 'Post 1',
        content: 'Lorem ipsum dolor sit amet 4',
        author: 'as61389dadhga62343hads6712',
        readers: [ 'as61389dadhga62343hads6712', '167asdf3689348sdad7312131s' ],
        createdAt: 1480793101559,
        _include: [ 'authorInfo', 'commentsInfo', 'readersInfo' ],
        _elapsed: { authorInfo: 2, commentsInfo: 2, readersInfo: 2, total: 2 },
        authorInfo: {
          id: 'as61389dadhga62343hads6712',
          name: 'Author 1',
          email: 'author1@posties.com',
          password: '2347wjkadhad8y7t2eeiudhd98eu2rygr',
          age: 55
        },
        commentsInfo:
          [
            { title: 'Comment 1',
              content: 'Lorem ipsum dolor sit amet 1',
              postId: 1 },
            { title: 'Comment 3',
              content: 'Lorem ipsum dolor sit amet 3',
              postId: 1 }
          ],
        readersInfo:
          [
            { id: 'as61389dadhga62343hads6712',
              name: 'Author 1',
              email: 'author1@posties.com',
              password: '2347wjkadhad8y7t2eeiudhd98eu2rygr',
              age: 55 },
            { id: '167asdf3689348sdad7312131s',
              name: 'Author 2',
              email: 'author2@posties.com',
              password: '2347wjkadhad8y7t2eeiudhd98eu2rygr',
              age: 16 }
          ]
      }
    };
    console.log('typeof schema1 before 1-st call', typeof schema1);
    
    for (let i = 1; i < 3; i++) {
      serialize(schema1)(hook);
      console.log(`typeof schema1 after ${i}-th call`, typeof schema1);
    }
  });
```
which produced
```
typeof schema1 before 1-st call function
serialize hook called. typeof schema function
shouldBe called
typeof schema1 after 1-th call function
serialize hook called. typeof schema function
shouldBe called
typeof schema1 after 2-th call function
    ✓ schema param is not changed
```
Can calling within a service change the conditions ?!?

However modifying function params has lead to problems several times, so I modified the code in https://github.com/feathersjs/feathers-hooks-common/pull/173 which will be published as v3.2.1.

**Please reopen if you can reproduce the problem with these changes.** Thanks.
